File: sdk-source/dart/example/quickstart.dart
// Runnable sample: `dart pub get && dart run example/quickstart.dart`
import 'dart:io';

import 'package:aegis/aegis.dart';

String envOr(String key, String fallback) {
  final value = Platform.environment[key];
  return (value == null || value.isEmpty) ? fallback : value;
}

Future<void> main() async {
  final aegis = Aegis(AegisOptions(
    baseUrl: envOr('AEGIS_BASE_URL', 'http://localhost:8080'),
    appKey: envOr('AEGIS_APP_KEY', ''),
    version: '1.0.0',
  ));

  try {
    final info = await aegis.init();
    stdout.writeln('initialized: ${info['status'] ?? 'ok'}');

    final version = info['version'] as Map<String, dynamic>?;
    if (version?['update_required'] == true) {
      stdout.writeln('mandatory update: ${version?['latest']}');
      return;
    }

    final auth = await aegis.login(
      username: envOr('AEGIS_USERNAME', 'demo'),
      password: envOr('AEGIS_PASSWORD', 'demo-password'),
    );
    stdout.writeln('signed in as ${(auth['user'] as Map?)?['username']}');

    await aegis.setVariable('last_seen', DateTime.now().toUtc().toIso8601String());
    aegis.startHeartbeat(onRevoked: (error) => stdout.writeln('session ended: ${error.message}'));

    stdout.writeln('authenticated: ${await aegis.isAuthenticated()}');
    await aegis.logout();
    stdout.writeln('signed out.');
  } on AegisException catch (error) {
    stderr.writeln(error);
    exitCode = 1;
  } finally {
    aegis.close();
  }
}

File: sdk-source/kotlin/src/main/kotlin/io/aegis/sdk/examples/Quickstart.kt
package io.aegis.sdk.examples

import io.aegis.sdk.Aegis
import io.aegis.sdk.AegisException
import io.aegis.sdk.AegisOptions
import io.aegis.sdk.bool
import io.aegis.sdk.obj
import io.aegis.sdk.str
import java.time.Duration

/** Runnable sample: `./gradlew run` or execute this main from your IDE. */
fun main() {
    val aegis = Aegis(
        AegisOptions(
            baseUrl = System.getenv("AEGIS_BASE_URL") ?: "http://localhost:8080",
            appKey = System.getenv("AEGIS_APP_KEY") ?: "",
            version = "1.0.0",
        ),
    )

    aegis.use {
        try {
            val info = it.init()
            println("initialized: ${info.str("status", "ok")}")

            if (info.obj("version").bool("update_required")) {
                println("mandatory update: ${info.obj("version").str("latest")}")
                return
            }

            val auth = it.login(
                System.getenv("AEGIS_USERNAME") ?: "demo",
                System.getenv("AEGIS_PASSWORD") ?: "demo-password",
            )
            println("signed in as ${auth.obj("user").str("username")}")

            it.setVariable("last_seen", java.time.Instant.now().toString())
            it.startHeartbeat(Duration.ofSeconds(60)) { reason -> println("session ended: $reason") }

            println("authenticated: ${it.isAuthenticated()}")
            it.logout()
            println("signed out.")
        } catch (error: AegisException) {
            System.err.println("Aegis error [${error.code}] ${error.message}")
            kotlin.system.exitProcess(1)
        }
    }
}